util/noeintr_close.c

```diff
@@ -0,0 +1,284 @@
+/**
+ * Goal: portably close a file descriptor in a multithreaded process without
+ * running into problems with EINTR.
+ *
+ * Run once:
+ * 1. Create a socket pair s[2].
+ * 2. Write a random cookie into s[0].
+ *
+ * To close(fd), wrap it in a mutex:
```

Comment:
s/wrap it in a mutex/wrapped in a mutex/ ?

Reply:
I think the tense is correct...?  The whole set of instructions are in present tense.

Arguably, the "wrap it in a mutex" should be part of the list of steps.  Maybe something like:
```
 * To close(fd),
 * 1. lock a mutex
 * 2. dup2(s[1], fd) until it succeeds.  (If it fails with !EINTR, bail.)
 * 3. close(fd).
 * 4. If we got EINTR, recv(fd, MSG_PEEK).
 * 5. If we read the random cookie, goto 3.
 * 6. unlock the mutex 
```

but I think that complicates it unnecessarily -- the "interesting" thing about those instructions are steps 2-5.  Adding steps for "lock & unlock" isn't going to clarify the idea.

Reply:
Well, it could be "wrap in a mutex", but the word "it" in "wrap it in a mutex" confuses me.  By "wrapped in a mutex..." I meant "do the following steps wrapped in a mutex".